### ISSUE TITLE
method to convert between different modes of FilePath

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7830]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#7830|
|Type|enhancement|
|Created|2015-03-24 09:53:16Z|
|Branch|https://github.com/twisted/twisted/tree/filepath-modeconvert-7830|

[#7805](https://github.com/twisted/twisted/issues/7805) implemented "modal" unicode support for `FilePath`, but it neglected to implement a mechanism to convert between text mode and bytes mode.

Any code which accesses the `path` or `sep` attribute, or calls a method like `basename` which doesn't take an argument, needs to expect a `FilePath` in a particular mode, so it gets those values in the correct type.  In order to expect a `FilePath` like this, there should be a way to convert and do nothing else.

We really want these methods to be on `FilePath` itself, so that we can implement smarter strategies in the future (for example, an undecodable path might still have a perfectly decodable `basename`, and if we preserve bytes internally and expose text externally we can save more information if it's a method on `FilePath` rather than some external construction mechanism).

These methods should return a new `FilePath` if a different mode is requested, since changing the mode on a shared `FilePath` could quite easily break things; however, since the only mutable state within a `FilePath` itself is cached stat information, it should be fine to return `self` in the case where the same mode is requested.

<details><summary>Searchable metadata</summary>

```
trac-id__7830 7830
type__enhancement enhancement
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__branches_filepath_modeconvert_7830 branches/filepath-modeconvert-7830
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1427190796808682 1427190796808682
changetime__1430102172968318 1430102172968318
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_1"></a>|@hawkowl commented|
|-|-|

(In [44448]) Branching to filepath-modeconvert-7830.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl removed owner|
|-|-|

Changes made, builders spun, please review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Changes looks good.

I guess that this will need another reviewer to check that the new method names make sense.

I am ok with them as I could not find a better name for them.

I don't see any tests exercising the option encoding argument ... if it is not needed maybe we can merge this without the encoding argument and add it later once an use case emerge otherwise, I think that we need tests to check that encoding argument is used.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_4"></a>|@hawkowl removed owner|
|-|-|

Added some tests for the encoding argument.

Builders spun, please review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban commented|
|-|-|

Changes looks good. Thanks!

I will leave it in the review queue so that the API is approved by another dev.

In case there is another work item depending on this, feel free to merge the current branch and continue the work.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_6"></a>|@hawkowl set status to `closed`|
|-|-|

(In [44576]) Merge filepath-modeconvert-7830: Methods to convert FilePaths to text/bytes modes

Author: hawkowl
Reviewer: adiroiban
Fixes: [#7830](https://github.com/twisted/twisted/issues/7830)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_7"></a>|@hawkowl commented|
|-|-|

Since it lines up with the existing API + terminology glyph proposed, I think it should be okay.
